docs/architecture-decisions/2023-01-11_beaker_productionisation_review.md

```diff
@@ -0,0 +1,409 @@
+
+## Findings summary
+
+The Beaker codebase is well factored and had a decent initial test coverage (albeit some of that test coverage is via a series of examples that while they provide high code coverage, don't actually validate all of the functionality).
+
+A series of changes have been landed to improve some of the fundamentals of Beaker in preparation for production launch:
+
+- [Various improvements](https://github.com/algorand-devrel/beaker/pull/142) - Improved test coverage, improved dev experience (setup + ongoing) via Poetry, improvements to the code quality setup (linting, automatic formatting, typing), allowed Windows development on Beaker itself, significantly improved CI/CD pipeline speed, removing the examples directory and tests from being distributed wit hthe PyPi package
+- [Typing improvements](https://github.com/algorand-devrel/beaker/pull/147)
+- [Removed inline imports](https://github.com/algorand-devrel/beaker/pull/148)
+- [Removed dead code](https://github.com/algorand-devrel/beaker/pull/149/files)
+- [Added automated release management and versioning](https://github.com/algorand-devrel/beaker/pull/161)
+
+In addition, there are a remaining set of more major (breaking) changes that are recommended. The recommendations are split into 2 categories, recommendations for immediate improvement (i.e. included in v1.0) and future suggestions that can be addressed post v1.0 launch.
+
+The recommended additional areas for immediate improvement are:
+
+- **Replace the class-based structure with an instance based one** - remove some areas of potential surprise for developers and simplify the Beaker codebase by moving to a composable instance-based structure rather than a static class-based structure
+- **Defer PyTEAL compilation** - improve flexibility and future contract output stability by deferring PyTEAL compilation (i.e. Beaker -> TEAL transpilation) to not happen when the Beaker contract is initialised
+- **Renamings** - There are some clear parameters that make sense to rename for various reasons
+- **Key decorator improvements** - Refactor some of the Beaker decorators to fix some bugs and improve user experience
+- **Beaker state refactor** - Refactor of the Beaker state interfaces to improve user extensibility and significantly simplify the Beaker codebase to improve maintainability
+
+The recommended areas for future improvement are:
+
+- Typed client generation from app spec to improve deploy-time and run-time dev experience
+- `Tmpl` values in app spec so you can have type-safe deployment clients that substitute any template values reliably at contract deploy time
+- Refactor storage types (blob, reserved, etc.) to allow use of in-built Python types and operators (terser, more intuitive)
+- Box storage implementation improved to match local/global behaviour and also automatically delete itself on contract deletion
```

Comment:
Very interested to know how this would be implemented as you need box names even on a delete transaction.

Reply:
With boxes being dynamic, yes it's not possible to completely handle this. However, given the limited Application storage, if you explicitly name a Box as a construct in Beaker, it would be good for it to be initialised the same was as Application storage and torn down automatically on deletion. In this way, for storage requirements known ahead of time (at least, ahead of create/update time), you can effectively have unlimited Application storage. This also raises the point of "migrations" which is something @barnjamin was keen on as well, I believe.

Reply:
Hello, is this to say you want to convert an empty box into additional key/value application storage space?

Reply:
@bbroder-algo we should make it trivial for the user to treat them as such :)